txscript/pkscript_test.go

```diff
@@ -0,0 +1,390 @@
+				i, test.pkScript, err)
+		case err == nil && !test.valid:
+			t.Fatalf("test #%d: successfully parsed invalid pkScript=%x",
+				i, test.pkScript)
+		}
+
+		if !test.valid {
+			continue
+		}
+
+		if !bytes.Equal(pkScript.Script(), test.pkScript) {
+			t.Fatalf("test #%d: expected to re-derive pkScript=%x, "+
+				"got pkScript=%x", i, test.pkScript,
+				pkScript.Script())
+		}
+	}
+}
+
+// TestComputePkScript ensures that we can correctly re-derive an output's
+// pkScript by looking at the input's signature script/witness attempting to
+// spend it.
+func TestComputePkScript(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		sigScript []byte
+		witness   wire.TxWitness
+		class     ScriptClass
+		pkScript  []byte
+	}{
+		{
+			sigScript: nil,
+			witness:   nil,
+			class:     NonStandardTy,
+			pkScript:  nil,
+		},
+		// Valid P2PKH.
+		{
+			sigScript: []byte{
+				// OP_DATA_71,
+				0x47,
+				// <71-byte sig>
+				0x30, 0x44, 0x02, 0x20, 0x65, 0x92, 0xd8, 0x8e,
+				0x1d, 0x0a, 0x4a, 0x3c, 0xc5, 0x9f, 0x92, 0xae,
+				0xfe, 0x62, 0x54, 0x74, 0xa9, 0x4d, 0x13, 0xa5,
+				0x9f, 0x84, 0x97, 0x78, 0xfc, 0xe7, 0xdf, 0x4b,
+				0xe0, 0xc2, 0x28, 0xd8, 0x02, 0x20, 0x2d, 0xea,
+				0x36, 0x96, 0x19, 0x1f, 0xb7, 0x00, 0xc5, 0xa7,
+				0x7e, 0x22, 0xd9, 0xfb, 0x6b, 0x42, 0x67, 0x42,
+				0xa4, 0x2c, 0xac, 0xdb, 0x74, 0xa2, 0x7c, 0x43,
+				0xcd, 0x89, 0xa0, 0xf9, 0x44, 0x54, 0x01,
+				// OP_DATA_33
+				0x21,
+				// <33-byte compressed pubkey>
+				0x02, 0x7d, 0x56, 0x12, 0x09, 0x75, 0x31, 0xc2,
+				0x17, 0xfd, 0xd4, 0xd2, 0xe1, 0x7a, 0x35, 0x4b,
+				0x17, 0xf2, 0x7a, 0xef, 0x30, 0x9f, 0xb2, 0x7f,
+				0x1f, 0x1f, 0x7b, 0x73, 0x7d, 0x9a, 0x24, 0x49,
+				0x90,
+			},
+			witness: nil,
+			class:   PubKeyHashTy,
+			pkScript: []byte{
+				// OP_DUP
+				0x76,
+				// OP_HASH160
+				0xa9,
+				// OP_DATA_20
+				0x14,
+				// <20-byte pubkey hash>
+				0xf0, 0x7a, 0xb8, 0xce, 0x72, 0xda, 0x4e, 0x76,
+				0x0b, 0x74, 0x7d, 0x48, 0xd6, 0x65, 0xec, 0x96,
+				0xad, 0xf0, 0x24, 0xf5,
+				// OP_EQUALVERIFY
+				0x88,
+				// OP_CHECKSIG
+				0xac,
+			},
+		},
+		// Valid P2SH (NP2WPKH).
+		{
+			// Since this is a NP2WPKH output, the script is that of
+			// a v0 P2WPKH.
+			sigScript: []byte{
+				// OP_0
+				0x00,
+				// OP_DATA_20
+				0x14,
+				// <20-byte pubkey hash>
+				0x1d, 0x7c, 0xd6, 0xc7, 0x5c, 0x2e, 0x86, 0xf4,
+				0xcb, 0xf9, 0x8e, 0xae, 0xd2, 0x21, 0xb3, 0x0b,
+				0xd9, 0xa0, 0xb9, 0x28,
+			},
+			// There is a witness NP2WPKH, but it is not needed to
+			// re-derive the pkScript.
+			witness: nil,
+			class:   ScriptHashTy,
+			pkScript: []byte{
+				// OP_HASH160
+				0xa9,
+				// OP_DATA_20
+				0x14,
+				// <20-byte script hash>
+				0x90, 0x1c, 0x86, 0x94, 0xc0, 0x3f, 0xaf, 0xd5,
+				0x52, 0x28, 0x10, 0xe0, 0x33, 0x0f, 0x26, 0xe6,
+				0x7a, 0x85, 0x33, 0xcd,
+				// OP_EQUAL
+				0x87,
+			},
+		},
+		// Valid P2SH.
+		{
+			sigScript: []byte{0x6b, 0x65, 0x6b}, // kek
```

Comment:
This isn't a valid sigScript for a P2SH output. It _must_ be canonical data push of the redeem script, and any other input script data. If this is updated to be an actual P2SH sigScript (can look on smartbit for examples), this test will fail.

Reply:
Fixed.